In [65]:
import pandas as pd
df = pd.read_csv('laboratory_specimen_project/raw_laboratory_specimens.csv')

In [66]:
df.head()

,specimen_id,collection_date,age_group,sex,department,specimen_type,test_requested,collection_shift,collector_role,rejection_status,rejection_reason,turnaround_time_min
0,SP101561,2025-01-22,0-9,Male,Emergency,Blood,Urea,Afternoon,Clinical Officer,Accepted,Not rejected,76
1,SP102836,2025-03-19,0-9,Female,Pediatrics,Blood,LFT,Morning,Phlebotomist,Accepted,Not rejected,38
2,SP104071,2025-09-23,50-59,Male,Pediatrics,Urine,Urine Culture,Morning,Phlebotomist,Accepted,Not rejected,111
3,SP101627,2025-06-17,0-9,Male,Maternity,Blood,Urea,Afternoon,Laboratory Technologist,Accepted,Not rejected,85
4,SP103306,2025-10-09,20-29,Female,Inpatient,Blood,Urea,Afternoon,Nurse,Rejected,Hemolysis,108


In [67]:
df.shape


(5018, 12)

In [68]:
df.columns

Index(['specimen_id', 'collection_date', 'age_group', 'sex', 'department',
       'specimen_type', 'test_requested', 'collection_shift', 'collector_role',
       'rejection_status', 'rejection_reason', 'turnaround_time_min'],
      dtype='object')

In [69]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5018 entries, 0 to 5017
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   specimen_id          5018 non-null   object
 1   collection_date      5018 non-null   object
 2   age_group            4993 non-null   object
 3   sex                  5018 non-null   object
 4   department           4978 non-null   object
 5   specimen_type        5018 non-null   object
 6   test_requested       5018 non-null   object
 7   collection_shift     5018 non-null   object
 8   collector_role       4988 non-null   object
 9   rejection_status     5018 non-null   object
 10  rejection_reason     5018 non-null   object
 11  turnaround_time_min  5018 non-null   int64 
dtypes: int64(1), object(11)
memory usage: 470.6+ KB


In [70]:
df.isnull().sum()

specimen_id             0
collection_date         0
age_group              25
sex                     0
department             40
specimen_type           0
test_requested          0
collection_shift        0
collector_role         30
rejection_status        0
rejection_reason        0
turnaround_time_min     0
dtype: int64

In [71]:
df.duplicated().sum()

np.int64(18)

In [72]:
df['department'].value_counts(dropna=False)

department
Outpatient     1525
Inpatient      1033
Emergency       793
Pediatrics      789
Maternity       411
ICU             407
NaN              40
 Outpatient      12
emergency         8
Name: count, dtype: int64

In [73]:
df['sex'].value_counts(dropna=False)

sex
Female    2548
Male      2470
Name: count, dtype: int64

In [74]:
df["age_group"].value_counts(dropna=False)

age_group
20-29    746
50-59    741
10-19    730
60+      721
40-49    709
0-9      678
30-39    668
NaN       25
Name: count, dtype: int64

In [75]:
df["specimen_type"].value_counts(dropna=False)

specimen_type
Blood      2019
Urine      1226
Swab        656
Sputum      560
Stool       300
CSF         210
 blood       26
URINE        14
sputum        7
Name: count, dtype: int64

In [76]:
df["test_requested"].value_counts(dropna=False)

test_requested
Urine Culture     623
Urinalysis        617
CBC               443
LFT               443
Culture           427
Urea              396
Creatinine        384
Glucose           379
Wound Culture     350
Throat Culture    306
AFB               196
Gram Stain        194
Stool O&P         153
CSF Analysis      107
Name: count, dtype: int64

In [77]:
df["collection_shift"].value_counts(dropna=False)

collection_shift
Morning      2232
Afternoon    1752
Night        1034
Name: count, dtype: int64

In [78]:
df["collector_role"].value_counts(dropna=False)

collector_role
Nurse                      2107
Clinical Officer           1265
Phlebotomist               1004
Laboratory Technologist     612
NaN                          30
Name: count, dtype: int64

In [79]:
df["rejection_status"].value_counts(dropna=False)

rejection_status
Accepted    4559
Rejected     459
Name: count, dtype: int64

In [80]:
df["rejection_reason"].value_counts(dropna=False)

rejection_reason
Not rejected           4559
Wrong container          75
Leaking container        72
Delayed transport        70
Insufficient volume      67
Clotted specimen         60
Mislabeling              58
Hemolysis                57
Name: count, dtype: int64

In [81]:
df.describe()

,turnaround_time_min
count,5018.000000
mean,88.486648
std,371.314729
min,-15.000000
25%,58.000000
50%,75.000000
75%,92.000000
max,9999.000000


In [82]:
df["turnaround_time_min"].describe()

count    5018.000000
mean       88.486648
std       371.314729
min       -15.000000
25%        58.000000
50%        75.000000
75%        92.000000
max      9999.000000
Name: turnaround_time_min, dtype: float64

In [83]:
df[['age_group', 'department', 'collector_role']].isnull().sum()

age_group         25
department        40
collector_role    30
dtype: int64

In [84]:
(df[['age_group', 'department', 'collector_role']].isnull().mean()*100).round(2)

age_group         0.5
department        0.8
collector_role    0.6
dtype: float64

In [85]:
df.duplicated().sum()

np.int64(18)

In [86]:
df[df.duplicated()]

,specimen_id,collection_date,age_group,sex,department,specimen_type,test_requested,collection_shift,collector_role,rejection_status,rejection_reason,turnaround_time_min
338,SP102961,2025-05-11,0-9,Female,Inpatient,Blood,LFT,Night,Phlebotomist,Accepted,Not rejected,66
1654,SP103235,2025-08-15,0-9,Male,Pediatrics,CSF,Culture,Morning,Clinical Officer,Accepted,Not rejected,86
1711,SP100615,2025-03-30,0-9,Male,ICU,Blood,Urea,Morning,Nurse,Accepted,Not rejected,73
1948,SP103624,2025-06-03,50-59,Female,Pediatrics,Urine,Urine Culture,Morning,Nurse,Accepted,Not rejected,48
2518,SP103708,2025-08-31,10-19,Female,Emergency,Blood,Glucose,Morning,Laboratory Technologist,Rejected,Leaking container,91
2549,SP104544,2025-09-21,60+,Male,Inpatient,Blood,Urea,Afternoon,Nurse,Accepted,Not rejected,93
2658,SP103335,2025-11-01,50-59,Female,Emergency,Swab,Wound Culture,Afternoon,Nurse,Accepted,Not rejected,30
3650,SP103406,2025-01-13,20-29,Male,Outpatient,Blood,CBC,Afternoon,Clinical Officer,Accepted,Not rejected,68
3702,SP103484,2025-12-05,0-9,Female,Inpatient,Blood,CBC,Afternoon,Nurse,Accepted,Not rejected,70
3735,SP102827,2025-09-25,10-19,Male,Pediatrics,Blood,Urea,Night,Nurse,Accepted,Not rejected,61


In [87]:
df[df.duplicated(keep=False)].sort_values("specimen_id")

,specimen_id,collection_date,age_group,sex,department,specimen_type,test_requested,collection_shift,collector_role,rejection_status,rejection_reason,turnaround_time_min
125,SP100615,2025-03-30,0-9,Male,ICU,Blood,Urea,Morning,Nurse,Accepted,Not rejected,73
1711,SP100615,2025-03-30,0-9,Male,ICU,Blood,Urea,Morning,Nurse,Accepted,Not rejected,73
4017,SP100640,2025-07-02,30-39,Female,Pediatrics,blood,Urea,Night,Nurse,Accepted,Not rejected,29
2611,SP100640,2025-07-02,30-39,Female,Pediatrics,blood,Urea,Night,Nurse,Accepted,Not rejected,29
3060,SP100757,2025-12-08,10-19,Male,Maternity,CSF,Culture,Morning,Laboratory Technologist,Accepted,Not rejected,122
4701,SP100757,2025-12-08,10-19,Male,Maternity,CSF,Culture,Morning,Laboratory Technologist,Accepted,Not rejected,122
4658,SP101868,2025-07-07,30-39,Female,Pediatrics,CSF,Culture,Night,Clinical Officer,Accepted,Not rejected,94
4387,SP101868,2025-07-07,30-39,Female,Pediatrics,CSF,Culture,Night,Clinical Officer,Accepted,Not rejected,94
2975,SP102827,2025-09-25,10-19,Male,Pediatrics,Blood,Urea,Night,Nurse,Accepted,Not rejected,61
3735,SP102827,2025-09-25,10-19,Male,Pediatrics,Blood,Urea,Night,Nurse,Accepted,Not rejected,61


In [88]:
df["specimen_id"].duplicated().sum()

np.int64(18)

In [89]:
df["specimen_id"].nunique()

5000

In [90]:
df[df["turnaround_time_min"] < 0]

,specimen_id,collection_date,age_group,sex,department,specimen_type,test_requested,collection_shift,collector_role,rejection_status,rejection_reason,turnaround_time_min
3316,SP104885,2025-03-11,0-9,Male,Emergency,Urine,Urinalysis,Morning,Nurse,Accepted,Not rejected,-15
4791,SP100095,2025-02-10,50-59,Male,Pediatrics,Blood,CBC,Morning,Nurse,Accepted,Not rejected,-15
4890,SP101374,2025-11-19,30-39,Female,ICU,Blood,CBC,Afternoon,Clinical Officer,Accepted,Not rejected,-15


In [91]:
df[df["turnaround_time_min"] > 500]

,specimen_id,collection_date,age_group,sex,department,specimen_type,test_requested,collection_shift,collector_role,rejection_status,rejection_reason,turnaround_time_min
192,SP103195,2025-07-21,10-19,Male,ICU,Blood,Glucose,Night,Nurse,Accepted,Not rejected,9999
1751,SP103263,2025-04-24,60+,Female,Outpatient,Stool,Culture,Morning,Nurse,Accepted,Not rejected,9999
2362,SP104066,2025-07-05,30-39,Female,Outpatient,Stool,Culture,Afternoon,Phlebotomist,Accepted,Not rejected,9999
2674,SP100561,2025-09-04,50-59,Female,ICU,CSF,Culture,Afternoon,Clinical Officer,Accepted,Not rejected,9999
3492,SP101608,2025-05-10,10-19,Female,Maternity,Urine,Urine Culture,Afternoon,Phlebotomist,Accepted,Not rejected,9999
4959,SP100619,2025-03-23,40-49,Female,Pediatrics,Blood,Glucose,Afternoon,Nurse,Accepted,Not rejected,9999
4998,SP101765,2025-05-23,20-29,Male,Outpatient,Swab,Wound Culture,Morning,Clinical Officer,Accepted,Not rejected,9999


In [92]:
df.loc[df["turnaround_time_min"].idxmin()]

specimen_id                SP104885
collection_date          2025-03-11
age_group                       0-9
sex                            Male
department                Emergency
specimen_type                 Urine
test_requested           Urinalysis
collection_shift            Morning
collector_role                Nurse
rejection_status           Accepted
rejection_reason       Not rejected
turnaround_time_min             -15
Name: 3316, dtype: object

In [93]:
df.loc[df["turnaround_time_min"].idxmax()]

specimen_id                SP103195
collection_date          2025-07-21
age_group                     10-19
sex                            Male
department                      ICU
specimen_type                 Blood
test_requested              Glucose
collection_shift              Night
collector_role                Nurse
rejection_status           Accepted
rejection_reason       Not rejected
turnaround_time_min            9999
Name: 192, dtype: object

In [94]:
import numpy as np

In [ ]:
df.loc[
    (df["turnaround_time_min"] < 0) | (df["turnaround_time_min"] > 500),
    "turnaround_time_min"
] = np.nan

In [ ]:
df["turnaround_time_min"].isnull().sum()

In [ ]:
df['turnaround_time_min'].describe()

In [ ]:
df["turnaround_time_min"].isnull().sum()

In [ ]:
df_raw = df.copy()

In [ ]:
print(df_raw.shape, df.shape)

In [ ]:
dept_map = {
    'outpatient': 'Outpatient',
    'inpatient': 'Inpatient',
    'emergency': 'Emergency',
    'pediatrics': 'Pediatrics',
    'maternity': 'Maternity',
    'icu': 'ICU'
}
df['department'] = df['department'].str.strip().str.lower().map(dept_map)

In [ ]:
df['department'].value_counts(dropna=False)

In [ ]:
specimen_map = {
    'blood': 'Blood',
    'urine': 'Urine',
    'swab': 'Swab',
    'sputum': 'Sputum',
    'stool': 'Stool',
    'csf': 'CSF'
}
df['specimen_type'] = df['specimen_type'].str.strip().str.lower().map(specimen_map)

In [ ]:
df['specimen_type'].value_counts(dropna=False)

In [ ]:
df.duplicated().sum()

In [ ]:
df[df.duplicated(keep=False)]

In [ ]:
df.duplicated(subset=['specimen_id']).sum()

In [ ]:
df = df.drop_duplicates(subset=["specimen_id"], keep="first")
df.shape

   ## Cleaning log
   - 18 rows were exact duplicates by specimen_id — removed, keeping first occurrence. Rows: 5018 → 5000.

In [ ]:
pd.crosstab(df['department'], df['sex'])

In [ ]:
df[(df['department'] == 'Maternity') & (df['sex'] == 'Male')].head(10)

In [ ]:
df['flag_dept_sex_mismatch'] = (df['department'] == 'Maternity') & (df['sex'] == 'Male')

179 rows show Maternity department with Male sex. Investigated for a common cause (collector, shift, date, specimen type) — none found, suggesting random data error rather than a systematic one. Flagged via flag_dept_sex_mismatch rather than deleted or corrected, since neither field can be confidently identified as the error.

In [ ]:
pd.crosstab(df['specimen_type'], df['test_requested'])

Checked specimen_type x test_requested via crosstab. All combinations are clinically valid(eg AFB only on sputum, Urine Culture only on urine). NO mismatches found.

In [ ]:
pd.crosstab(df['department'], df['age_group'])

In [ ]:
df[(df['department'] == 'Pediatrics') & (~df['age_group'].isin(['0-9', '10-19']))].head(10)

In [ ]:
mismatch = df[(df["department"] == "Pediatrics") & (~df["age_group"].isin(["0-9", "10-19"]))]
print(mismatch.shape)
print(mismatch["collector_role"].value_counts(normalize=True))
print(mismatch["collection_shift"].value_counts(normalize=True))

In [ ]:
print(df["collector_role"].value_counts(normalize=True))
print(df["collection_shift"].value_counts(normalize=True))

In [ ]:
df["flag_dept_age_mismatch"] = (df["department"] == "Pediatrics") & (~df["age_group"].isin(["0-9", "10-19"]))
df["flag_dept_age_mismatch"].sum()  # should be 568

568 rows show Pediatrics department with adult age_group(20-29 through 60+). Checked for concentration by collector_roles and collection_shifts- proportions match the overall dataset within ~2points, so no traceable cause found. Flagged via flag_dept_age_mismatch, consitent with the Marternity/sex mismatch treatment.

In [ ]:
df[df['department'].isnull()]['collector_role'].value_counts(normalize=True)
df[df['department'].isnull()]['collection_shift'].value_counts(normalize=True)

In [ ]:
df['department'] = df['department'].fillna('Unknown')
df['collector_role'] = df['collector_role'].fillna('Unknown')
df['age_group'] = df['age_group'].fillna('Unknown')

In [ ]:
print(df["department"].isnull().sum())
print(df["age_group"].isnull().sum())
print(df["collector_role"].isnull().sum())
print(df["department"].value_counts())
print(df["age_group"].value_counts())
print(df["collector_role"].value_counts())

In [ ]:
df.to_csv('specimens_clean.csv', index=False)

In [ ]:
import os
print(os.getcwd())
print(os.path.exists('specimens_clean.csv'))

In [ ]:
from sqlalchemy import create_engine, text

# adjust user/password/host if yours differ
engine = create_engine("mysql+pymysql://labuser:your_password_here@localhost:3306/")

with engine.connect() as conn:
    result = conn.execute(text("SHOW DATABASES;"))
    for row in result:
        print(row)

In [ ]:
with engine.connect() as conn:
    conn.execute(text("CREATE DATABASE IF NOT EXISTS lab_specimens;"))

In [ ]:
engine = create_engine("mysql+pymysql://labuser:analyst123@localhost:3306/lab_specimens")

In [ ]:
create_sql = """
CREATE TABLE IF NOT EXISTS specimens (
    specimen_id            VARCHAR(10) PRIMARY KEY,
    collection_date        DATE,
    age_group              VARCHAR(10),
    sex                    VARCHAR(10),
    department             VARCHAR(20),
    specimen_type          VARCHAR(20),
    test_requested         VARCHAR(30),
    collection_shift       VARCHAR(15),
    collector_role         VARCHAR(30),
    rejection_status       VARCHAR(15),
    rejection_reason       VARCHAR(30),
    turnaround_time_min    FLOAT NULL,
    flag_dept_sex_mismatch BOOLEAN,
    flag_dept_age_mismatch BOOLEAN
);
"""

with engine.connect() as conn:
    conn.execute(text(create_sql))

In [ ]:
with engine.connect() as conn:
    for row in conn.execute(text("DESCRIBE specimens;")):
        print(row)

In [ ]:
df.to_sql("specimens", con=engine, if_exists="append", index=False)

In [ ]:
with engine.connect() as conn:
    print(conn.execute(text("SELECT COUNT(*) FROM specimens;")).fetchone())
    print(conn.execute(text("SELECT COUNT(*) FROM specimens WHERE turnaround_time_min IS NULL;")).fetchone())

In [ ]:
engine = create_engine("mysql+pymysql://labuser:your_password_here@localhost:3306/lab_specimens")